tenpy/simulations/time_evolution.py

```diff
@@ -74,3 +82,411 @@ def final_measurements(self):
+            if 'resume_data_bra' in self.results['resume_data']:
+                self.logger.info("use `resume_data` for initializing the algorithm engine")
+                resume_data_bra = self.results['resume_data']['resume_data_bra'].copy()
+                # clean up: they are no longer up to date after algorithm initialization!
+                # up to date resume_data is added in :meth:`prepare_results_for_save`
+                self.results['resume_data']['resume_data_bra'].clear()
+                del self.results['resume_data']['resume_data_bra']
+
+        super().init_algorithm(**kwargs)  # links to Simulation
+        if resume_data_bra is not None:
+            kwargs.setdefault('resume_data', resume_data_bra)
+            if 'psi' in resume_data_bra:
+                self.psi_ground_state = resume_data_bra['psi']  # make sure to use resume data of bra
+        kwargs.setdefault('cache', self.cache)  # TODO: can we use the same cache
+        # make sure a second engine is used when evolving the bra
+        # fetch engine that evolves ket
+        AlgorithmClass = self.engine.__class__
+        # instantiate the second engine for the ground state
+        algorithm_params = self.options.subconfig('algorithm_params')
+        self.engine_bra = AlgorithmClass(self.psi_ground_state, self.model,
+                                         algorithm_params, **kwargs)
+
+    def run_algorithm(self):
+        while True:
+            if np.real(self.engine.evolved_time) >= self.final_time:
+                break
+            self.logger.info("evolve to time %.2f, max chi=%d", self.engine.evolved_time.real,
+                             max(self.psi.chi))
+            self.engine_bra.run()  # first evolve bra
+            # call engine_bra_resume_data in case something else is done here....
+            self.engine.run()  # evolve ket (psi)
+            # sanity check, bra and ket should evolve to same time
+            assert np.isclose(self.engine_bra.evolved_time, self.engine.evolved_time), ('Bra evolved to different time '
+                                                                                        'than ket')
+            self.model = self.engine.model
+            self.make_measurements()
+            self.engine.checkpoint.emit(self.engine)  # set up in init_algorithm of Simulation class
+            # and connects to self.save_at_checkpoint which ultimately calls self.save_results()
+            # if there are no results self.prepare_results_for_save() which calls get_resume_data() which
+            # calls engine.get_resume_data
+
+    def m_correlation_function(self, results, psi, model, simulation, **kwargs):
+        """Equivalent to :meth:`TimeDependentCorrelation.m_correlation_function`."""
+        self.logger.info("calling m_correlation_function")
+        operator_t = to_iterable(self.operator_t)
+        psi_bra = self.engine_bra.psi.copy()  # make copy since algorithm might use grouped bra
```

Comment:
This is a small optimization, but you can move the copy to the if self.grouped > 1: case which is the only one where youneed it, isn't it?